invokeai/backend/stable_diffusion/extensions_manager.py

```diff
@@ -0,0 +1,167 @@
+        pass
+
+    @abstractmethod
+    def post_unet(self, ctx: DenoiseContext, ext_manager: ExtensionsManager):
+        pass
+
+    @abstractmethod
+    def post_apply_cfg(self, ctx: DenoiseContext, ext_manager: ExtensionsManager):
+        pass
+
+
+class ProxyCallsClass:
+    def __init__(self, handler):
+        self._handler = handler
+
+    def __getattr__(self, item):
+        return partial(self._handler, item)
+
+
+class CallbackInjectionPoint:
+    def __init__(self):
+        self.handlers = {}
+
+    def add(self, func: Callable, order: int):
+        if order not in self.handlers:
+            self.handlers[order] = []
+        self.handlers[order].append(func)
+
+    def __call__(self, *args, **kwargs):
+        for order in sorted(self.handlers.keys(), reverse=True):
+            for handler in self.handlers[order]:
+                handler(*args, **kwargs)
+
+
+class ExtensionsManager:
+    def __init__(self):
+        self.extensions = []
+
+        self._callbacks = {}
+        self.callbacks: ExtCallbacksApi = ProxyCallsClass(self.call_callback)
+
+    def add_extension(self, ext: ExtensionBase):
+        self.extensions.append(ext)
+
+        self._callbacks.clear()
+
+        for ext in self.extensions:
+            for inj_info in ext.injections:
+                if inj_info.type == "callback":
+                    if inj_info.name not in self._callbacks:
+                        self._callbacks[inj_info.name] = CallbackInjectionPoint()
+                    self._callbacks[inj_info.name].add(inj_info.function, inj_info.order)
+
+                else:
+                    raise Exception(f"Unsupported injection type: {inj_info.type}")
+
+    def call_callback(self, name: str, *args, **kwargs):
+        if name in self._callbacks:
+            self._callbacks[name](*args, **kwargs)
+
+    # TODO: is there any need in such high abstarction
+    # @contextmanager
+    # def patch_extensions(self):
+    #    exit_stack = ExitStack()
+    #    try:
+    #        for ext in self.extensions:
+    #            exit_stack.enter_context(ext.patch_extension(self))
+    #
+    #        yield None
+    #
+    #    finally:
+    #        exit_stack.close()
+
+    @contextmanager
+    def patch_attention_processor(self, unet: UNet2DConditionModel, attn_processor_cls: object):
+        unet_orig_processors = unet.attn_processors
+        exit_stack = ExitStack()
+        try:
+            # just to be sure that attentions have not same processor instance
+            attn_procs = {}
+            for name in unet.attn_processors.keys():
+                attn_procs[name] = attn_processor_cls()
+            unet.set_attn_processor(attn_procs)
+
+            for ext in self.extensions:
+                exit_stack.enter_context(ext.patch_attention_processor(attn_processor_cls))
+
+            yield None
+
+        finally:
+            unet.set_attn_processor(unet_orig_processors)
+            exit_stack.close()
```

Comment:
It can be separated, but I done this way, so that all attention processor logic in same place. It's more like "initialize environment".
And patching unet(it's structure and weights) all in `patch_unet`

Reply:
I wrote this code as I see it. Now, probably should be created new comment as this linked to outdated code.